feature/text_based.py

```diff
@@ -0,0 +1,694 @@
+        """
+        kmeans = KMeans(n_clusters=selected_size, random_state=self.seed, n_init=self.trials)
+        distances = kmeans.fit_transform(self.features)
+
+        ### NEXT_S_DIS
+        for i in range(self._num_cols):
+            if np.min(distances[i]) == 0:
+                nonzero_distances = distances[i][np.nonzero(distances[i])]
+                if len(nonzero_distances) > 0:
+                    min_nonzero_distance = np.min(nonzero_distances)
+                    distances[i][distances[i] == 0] = min_nonzero_distance
+        ###
+
+        diversity_cost = [np.min(distances[i]) for i in range(self._num_cols)]
+
+        ### DUMMY
+        # diversity_cost = [c + 1e-1 for c in diversity_cost]
+        ###
+
+        # Scale contexts so that sum of costs remain constant
+        diversity_cost = [c * self._num_cols / sum(diversity_cost) for c in diversity_cost]
+
+        return diversity_cost
+
+    def _get_num_row_covered(self, selected) -> int:
+        check_true(self.matrix is not None, "matrix is not initialized")
+
+        # Create indicator to count number of covered rows
+        row_covered = np.sum(self.matrix[:, selected], axis=1)
+
+        row_covered[row_covered > 0] = 1
+
+        num_row_covered = sum(row_covered)
+
+        return num_row_covered
+
+    def _get_selected_size(self) -> Tuple[np.ndarray, int]:
+        if self.cost_metric == "unicost":
+            unicost = np.ones(self._num_cols)
+            data = Data(cost=unicost, matrix=self.matrix)
+            unicost_selected = self._solve_set_cover(data)
+            selected_size = len(unicost_selected)
+        else:
+            diversity_cost = self._get_diversity_cost(self._num_cols)
+            data = Data(cost=diversity_cost, matrix=self.matrix)
+            diversity_selected = self._solve_set_cover(data)
+            selected_size = len(diversity_selected)
+            unicost = diversity_cost
+        return unicost, selected_size
+
+    def _get_closest_to_centroids(self, kmeans: KMeans) -> List:
+        df = pd.DataFrame({"cluster": kmeans.labels_})
+        for c in np.unique(kmeans.labels_):
+            # Get indices of cluster
+            mask = df["cluster"] == c
+
+            # Squared error to cluster centroid
+            dist = np.sum((self.features[mask] - kmeans.cluster_centers_[c]) ** 2, axis=1)
+
+            # Create column in df
+            df.loc[mask, "dist_to_centroid"] = dist
+
+        kmeans_selected_df = df.loc[df.groupby('cluster', sort=False).dist_to_centroid.idxmin()]
+        kmeans_selected = kmeans_selected_df.index.values.tolist()
+
+        return kmeans_selected
+
+    def _solve_max_cover(self, data: Data, selected: List) -> List:
+        # If selected is given, limit the max_cover_size
+        if selected is not None and self.selection_size is not None and len(selected) > 0:
+            assert (self.selection_size <= len(selected)), "Max Cover Error: max_cover_size cannot exceed num selected"
```

Comment:
remove assert, and make sure to use Exception

Reply:
actually I am not sure about this assert. say more why is this needed.

Reply:
The assert statement in the _solve_max_cover function was appropriate for the old structure of the ContentSelector class. However, in the new MaxCoverSelector class, this condition is already checked before calling the function, so the assert statement is redundant and has been removed in the new commit.